data/maps/SP_Scenarios/Along The River.wmf/scripting/functions.lua

```diff
@@ -0,0 +1,143 @@
+-- =======================================================================
+--                             Along The River
+-- =======================================================================
+
+-- Checks objective for done flag.
+function obj_done(objective)
+	local done = false
+	if objective then
+		done = objective.done
+	end
+	return done
+end
+
+-- =======================================================================
+
+-- Places building at field if player owns field and extra field for flag.
+function place_building_if_owner(player, building, point, cs)
+	local build = false
+	if point.owner == player and point.brn.owner == player then
+		player:place_building(building, point, cs, true)
+		build = true
+	end
+	return build
+end
+
+-- =======================================================================
+
+function remove_type(point, _type)
+	local immovable = point.immovable
+	if immovable then
+		if immovable.descr.type_name == _type then
+			immovable:remove()
+		end
+	end
+end
+
+-- =======================================================================
+
+function immovable_has_name(immovable, name)
+	local valid = false
+	if immovable then
+		if immovable.descr.name == name then valid = true end
+	end
+	return valid
+end
+
+-- =======================================================================
+
+function add_wares(point, ware, amount)
+	local immovable = point.immovable
+	if immovable then
+		if immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite" then
+			immovable:set_wares(ware, immovable:get_wares(ware) + amount)
+		end
+	end
+end
+
+-- =======================================================================
+
+-- Transports wares to flag, warehouse or productionsite.
+-- Waits if the building does not exist.
+function transport_wares(player, point, ware, amount, alert, image)
+	local count = amount
+	while count > 0 do
+		sleep(499)
+		local immovable = point.immovable
+		if immovable then
+			if immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite" then
+				immovable:set_wares(ware, immovable:get_wares(ware) + 1)
+				count = count - 1
+			end
+		end
+	end
+	if(alert) then
+      local ware_name =  wl.Game():get_ware_description(ware).descname
+		local immovable = point.immovable
+		if immovable then
+			if immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite" then
+				if image then
+					send_to_inbox(player, _"Transport", li_image(image, amount.._" units of "..ware_name.." have been brought to your "..immovable.descr.descname), { field = point, popup = true, })
+				else
+					send_to_inbox(player, _"Transport", p(amount.._" units of "..ware_name.." have been brought to your "..immovable.descr.descname), { field = point, popup = true, })
```

Comment:
Like this: amount .. _(" units of ") .. ware_name .. _("  ?

Reply:
On second thoughts… this should be an ngettext+bformat call with placeholders, like
```
ngettext(
   "%1% unit of %2% has been brought to your %3%.",
   "%1% units of %2% have been brought to your %3%.",
   amount):bformat(amount, ware_name, immovable.descr.descname)
```
I wonder if this could be rephrased in some way because the insertion of undeclined ware and building name would break the grammar in some languages.

Reply:
you are right with ngettext. will think about proper rephrasing.

Reply:
This still leaves the issue that the building name should inflect the article, e.g. low german
> Dien Raadhuus hett deese Waar kregen: 1× Flint
> Dien**e** Muu hett deese Waren kregen: 2× Reit

How about leaving out the building name from the sentence ("You have received %1% unit|units of the ware ‘%2%’.") and adding the building icon to the message and linking the message's Goto location to the building?

Reply:
maybe I found a solution